### PR TITLE
Add CIF Writer and remove PDBWriter.

### PR DESCRIPTION
**Pull request by proteneer**
This PR:

1) Removes the PDB limit of 99,999 atoms
2) Removes the PDB limit of 9,999 residues
3) Removes the need to generate a tempfile by directly merging rdkit mols into an OpenMM topology, as opposed to relying on Chem.CombineMols
4) No longer need to explicitly call `writer.close()`

### CODE REVIEW

File: timemachine/fe/cif_writer.py [hunk 1]
```diff
@@ -37,67 +34,84 @@ def convert_single_topology_mols(coords: np.ndarray, atom_map: AtomMapMixin) ->
...
+                # (ytz): while we could get fancier, mmcif only has the covale bond type:
+                # https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v40.dic/Items/_struct_conn_type.id.html
+                # so we only bother with "yes/no"
+                for bond in mol.GetBonds():
+                    src_atom = old_idx_to_new_atom_map[bond.GetBeginAtomIdx()]
+                    dst_atom = old_idx_to_new_atom_map[bond.GetEndAtomIdx()]
+                    combined_topology.addBond(src_atom, dst_atom)
+
+            else:
+                raise ValueError("Unknown obj type")
```
Comment (badisa) [inline timemachine/fe/cif_writer.py:109]: nit:
```suggestion
                raise ValueError(f"Unknown obj type: {type(obj)}")
```

File: timemachine/fe/cif_writer.py [hunk 2]
```diff
@@ -37,67 +34,84 @@ def convert_single_topology_mols(coords: np.ndarray, atom_map: AtomMapMixin) ->
...
+                new_residue = combined_topology.addResidue(name="LIG" + str(idx), chain=new_chain)
+                old_idx_to_new_atom_map = {}
+                for atom in mol.GetAtoms():
+                    name = atom.GetSymbol() + str(atom.GetIdx())
+                    element = app.element.Element.getByAtomicNumber(atom.GetAtomicNum())
+                    new_atom = combined_topology.addAtom(name, element, new_residue)
+                    old_idx = atom.GetIdx()
+                    old_idx_to_new_atom_map[old_idx] = new_atom
+
+                # (ytz): while we could get fancier, mmcif only has the covale bond type:
```
Comment (badisa) [inline timemachine/fe/cif_writer.py:100]: Looks like you are referring to the code, but reads like a typo. Not sure if this matters

```suggestion
                # (ytz): while we could get fancier, mmcif only has the covalent bond type:
```

File: tests/test_interpolate_fe.py [hunk 3]
```diff
@@ -63,12 +63,11 @@ def test_hif2a_free_energy_estimates():
         # continue
         frames = simulate_system(U_fn, x0, num_samples=2000)
         all_frames.append(frames)
-        writer = pdb_writer.PDBWriter([mol_a, mol_b], "debug_" + str(lambda_idx) + ".pdb")
+        writer = cif_writer.CIFWriter([mol_a, mol_b], "debug_" + str(lambda_idx) + ".pdb")
```
Comment (badisa) [inline tests/test_interpolate_fe.py:66]: nit:
```suggestion
        writer = cif_writer.CIFWriter([mol_a, mol_b], "debug_" + str(lambda_idx) + ".cif")
```

File: tests/test_chiral_restraints.py [hunk 4]
```diff
@@ -591,9 +591,9 @@ def assert_same_signs(a, b):
     x0_inverted = utils.get_romol_conf(mol_inverted)
 
     # don't remove - useful for visualization
-    # from timemachine.fe import pdb_writer
+    # from timemachine.fe import cif_writer
     # traj = minimize_scipy(U_fn, x0_inverted, return_traj=True)
-    # writer = pdb_writer.PDBWriter([mol], "chiral_inversion.pdb")
+    # writer = cif_writer.CIFWriter([mol], "chiral_inversion.pdb")
```
Comment (badisa) [inline tests/test_chiral_restraints.py:596]:
```suggestion
    # writer = cif_writer.CIFWriter([mol], "chiral_inversion.cif")
```

File: examples/relative_free_energy.py [hunk 5]
```diff
@@ -18,13 +18,12 @@ def write_trajectory_as_pdb(mol_a, mol_b, core, all_frames, host_topology, prefi
     atom_map_mixin = AtomMapMixin(mol_a, mol_b, core)
     for window_idx, window_frames in enumerate(all_frames):
         out_path = f"{prefix}_{window_idx}.pdb"
```
Comment (badisa) [inline examples/relative_free_energy.py:20]:
```suggestion
        out_path = f"{prefix}_{window_idx}.cif"
```

**Comment by badisa** [comment]
Would be good to also rename the following two functions to say `cif` rather than `pdb`

https://github.com/proteneer/timemachine/blob/master/tests/test_single_topology_confgen.py#L21
https://github.com/proteneer/timemachine/blob/master/examples/relative_free_energy.py#L16

File: timemachine/fe/cif_writer.py [hunk 6]
```diff
@@ -37,67 +34,84 @@ def convert_single_topology_mols(coords: np.ndarray, atom_map: AtomMapMixin) ->
...
+
         Usage:
 
         ```
         topol = app.PDBFile("1dfr.pdb").topology
         mol_a = Chem.MolFromMolBlock("...")
         mol_b = Chem.MolFromMolBlock("...")
 
-        writer = PDBWriter([topol, mol_a, mol_b], out.pdb) # writer header
+        writer = PDBxWriter([topol, mol_a, mol_b], out.pdb) # writer header
```
Comment (jkausrelay) [inline timemachine/fe/cif_writer.py:57]: CIFWriter

Comment (proteneer) [reply timemachine/fe/cif_writer.py:57]: willfix

File: timemachine/fe/cif_writer.py (hunk 2 above)
Comment (badisa) [reply timemachine/fe/cif_writer.py:100]: Maybe just adding quotes to indicate that it is intentional? Anyways, just a nit since I read it as a typo

Comment (proteneer) [reply timemachine/fe/cif_writer.py]: done

File: timemachine/fe/cif_writer.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+                    for old_atom in old_residue.atoms():
+                        new_atom = combined_topology.addAtom(old_atom.name, old_atom.element, new_residue)
+                        old_atom_id_kv[old_atom.id] = new_atom
+
+                for old_bond in old_topology.bonds():
+                    old_atom1_id = old_bond.atom1.id
+                    old_atom2_id = old_bond.atom2.id
+                    new_atom1 = old_atom_id_kv[old_atom1_id]
+                    new_atom2 = old_atom_id_kv[old_atom2_id]
+                    combined_topology.addBond(new_atom1, new_atom2, old_bond.type, old_bond.order)
```
Comment (jkausrelay) [inline timemachine/fe/cif_writer.py:124]: should this just call deep_copy_topology ?

File: timemachine/fe/cif_writer.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        """
+        Write a coordinate frame
+
+        Parameters
+        ----------
+        x: np.ndarray
+            coordinates in units of angstroms
+
+        """
+        self.frame_idx += 1
```
Comment (badisa) [inline timemachine/fe/cif_writer.py:166]: nit: To verify, the frame idx is 1 indexed? Not sure if it matters

File: timemachine/fe/cif_writer.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+                    old_idx = atom.GetIdx()
+                    old_idx_to_new_atom_map[old_idx] = new_atom
+
+                # (ytz): while we could get fancier, mmcif only has the 'covale' bond type:
+                # https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v40.dic/Items/_struct_conn_type.id.html
+                # so we only bother with "yes/no"
+                for bond in mol.GetBonds():
+                    src_atom = old_idx_to_new_atom_map[bond.GetBeginAtomIdx()]
+                    dst_atom = old_idx_to_new_atom_map[bond.GetEndAtomIdx()]
+                    combined_topology.addBond(src_atom, dst_atom)
```
Comment (jkausrelay) [inline timemachine/fe/cif_writer.py:144]: Does addBond support bond order ?

Comment (proteneer) [reply timemachine/fe/cif_writer.py:116]: it does but mmCIF doesn't.

**Review by badisa — APPROVED** [review]

File: timemachine/fe/cif_writer.py (hunk 8 above)
Comment (proteneer) [reply timemachine/fe/cif_writer.py:138]: yes - same as PDB, things like `HETATM` records are all 1-indexed

**Pull request merged**